crates/propolis-api-types/src/instance_spec/mod.rs

```diff
@@ -191,6 +191,13 @@ pub mod v0;
 // This type derives `SerializeDisplay` and `DeserializeFromStr` so that it can
 // be used as a map key when serializing to JSON, which requires strings (and
 // not objects) as keys.
+//
+// WARNING: Progenitor-generated versions of this type will use serde's default
+// derived Serialize and Deserialize implementations, not the serde_with
+// implementations. This can cause clients to serialize spec keys in a form that
+// the server can't deserialize. To get around this, Progenitor clients that
+// use spec keys should use a `replace` directive to replace the generated
+// type with this type.
```

Comment:
It occurs to me that there might be an alternative solution to this that doesn't require "WARNING: remember to do this extra stuff when generating clients": we are presently deriving `JsonSchema`, and I assume the generated schema is for a tagged enum representation. Could we, instead, add `serde(untagged)` and hand-implement `JsonSchema` for `SpecKey`, the way [we do for `NameOrId` in `omicron-common`](https://github.com/oxidecomputer/omicron/blob/aeb27f08c883c04360883fbb9ffd5f1222ceeede/common/src/api/external/mod.rs#L394-L414)? That way, I think we _should_ instead generate an OpenAPI spec that represents this as just a String but with hints suggesting it could be a UUID. I'm not 100% sure that will get us the desired behavior but it seems worth looking into.

Reply:
I am embarrassed to admit that I did not see the parallel between this and `NameOrId`! I'll give this a try.

Reply:
I was able to get this to work in fc7a85f. It didn't exactly spark joy, but it does get rid of the `serde_with` dependency and makes things slightly nicer for sled-agent, so I think I'm leaning toward going with it anyway. Would be interested to hear what you think!

Reply:
I'll take a look!